### PR TITLE
No LinkTimeCodeGeneration in VS

### PR DESCRIPTION
**Pull request by guolinke**
It seems LTCG will slow down the compile speed, while it does not improve the performance.

### CODE REVIEW

File: windows/LightGBM.vcxproj [hunk 1]
```diff
@@ -103,6 +103,8 @@
     </Link>
     <Link>
       <AdditionalDependencies>msmpi.lib</AdditionalDependencies>
+      <LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 2]
```diff
@@ -103,6 +103,8 @@
...
+      </LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 3]
```diff
@@ -122,6 +124,8 @@
     <Link>
       <AdditionalDependencies>
       </AdditionalDependencies>
+      <LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 4]
```diff
@@ -122,6 +124,8 @@
...
+      </LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 5]
```diff
@@ -150,6 +154,8 @@
       <EnableCOMDATFolding>true</EnableCOMDATFolding>
       <OptimizeReferences>true</OptimizeReferences>
       <AdditionalDependencies>msmpi.lib</AdditionalDependencies>
+      <LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 6]
```diff
@@ -150,6 +154,8 @@
...
+      </LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 7]
```diff
@@ -170,6 +176,8 @@
     <Link>
       <AdditionalDependencies />
       <OptimizeReferences>true</OptimizeReferences>
+      <LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 8]
```diff
@@ -170,6 +176,8 @@
...
+      </LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 9]
```diff
@@ -191,6 +199,8 @@
       <AdditionalDependencies>
       </AdditionalDependencies>
       <OptimizeReferences>true</OptimizeReferences>
+      <LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

File: windows/LightGBM.vcxproj [hunk 10]
```diff
@@ -191,6 +199,8 @@
...
+      </LinkTimeCodeGeneration>
```
Comment (guolinke) [inline windows/LightGBM.vcxproj]:
```suggestion
```

**Comment by StrikerRUS** [comment]
Sorry, I have never hear about `LTCG` option. But I've found related discussion: https://developercommunity.visualstudio.com/content/problem/501385/very-slow-linking-of-a-ccli-project-with-ltcg-enab.html

**Comment by guolinke** [comment]
@StrikerRUS `WholeProgramOptimization` will enable `ltcg` by default.

**Comment by guolinke** [comment]
close due to `WholeProgramOptimization` is silghtly faster..

**Comment by guolinke** [comment]
Benchmarks: 

| Runs      | with /GL|  no /GL |
|----|  ----| ----| 
| 1 | 9.69 s | 9.06 s |
| 2 | 9.13 s | 9.06 s|
| 3 | 9.35 s | 9.16 s|
| 4 | 9.24 s | 9.14 s|
| 5 | 9.07 s | 9.11 s|

**Comment by jameslamb** [comment]
Removing my requested review, I don't know anything about this.

**Pull request merged**